src/policy/mallocspace/global.rs

```diff
@@ -214,45 +249,180 @@ impl<VM: VMBinding> MallocSpace<VM> {
+        match VM::VMObjectModel::LOCAL_MARK_BIT_SPEC {
+            MetadataSpec::OnSide(local_mark_bit_side_spec) => {
+                unsafe { self.sweep_chunk_mark_on_side(chunk_start, local_mark_bit_side_spec) };
+            }
+            _ => {
+                unsafe { self.sweep_chunk_mark_in_header(chunk_start) };
+            }
+        }
+    }
 
-        // To sum up the total size of live objects. We check this against the active_bytes we maintain.
+    /// This function is called when the mark bits sit on the side metadata.
+    /// This has been optimized with the use of bulk loading and bulk zeroing of
+    /// metadata.
+    ///
+    /// # Safety
+    /// unsafe as it uses non-atomic accesses to side metadata (although these
+    /// non-atomic accesses should not have race conditions associated with them)
+    /// as well as calling libc functions
+    unsafe fn sweep_chunk_mark_on_side(
+        &self,
+        chunk_start: Address,
+        mark_bit_spec: SideMetadataSpec,
+    ) {
         #[cfg(debug_assertions)]
         let mut live_bytes = 0;
 
-        debug!(
-            "Used bytes before releasing: {}",
-            self.active_bytes.load(Ordering::Relaxed)
-        );
+        debug!("Check active chunk {:?}", chunk_start);
+        let mut chunk_is_empty = true;
+        let mut address = chunk_start;
+        let chunk_end = chunk_start + BYTES_IN_CHUNK;
+        let mut page = conversions::page_align_down(address);
+        let mut page_is_empty = true;
+        let mut on_page_boundary = false;
+        let align: usize = 128 * (1 << LOG_MIN_OBJECT_SIZE);
+
+        while address < chunk_end {
+            // We extensively tested the performance of the following if-statement and were
+            // surprised to note that in the case of newer AMD microarchitecures (>= Zen), some
+            // microarchitectural state/idiosyncrasies result in favourable cache placement/locality
+            // for the case where the conditionals (i.e. just the body of both the if-statements are left
+            // in the hot loop) which lead to a large performance speedup. Even more surprising was the
+            // revelation that the hot loop has worse cache placement/locality if the entire if-statement
+            // was commented out -- effectively meaning that [more work in the hot loop => better performance]
+            // which was counterintuitive to our beliefs.
+            //
+            // The performance tradeoffs on Intel and older AMD microarchitectures were as expected, i.e.
+            // wherein the performance of the hot loop decreased if more work was done in the loop.
+            if address - page >= BYTES_IN_PAGE {
+                if page_is_empty {
+                    unset_page_mark_unsafe(page);
+                }
+                page = conversions::page_align_down(address);
+                page_is_empty = !on_page_boundary;
```

Comment:
No that's not the case? `on_page_boundary` is set to false before right after this

Reply:
I meant in the line `page_is_empty = !on_page_boundary`. I don't get the logic here. Why `page_is_empty` is the opposite of `on_page_boundary`?

Reply:
If an object spans a page boundary, then the next page is not empty. Hence it's the opposite

Reply:
I think you're right and it should be >= for both then

Reply:
> If an object spans a page boundary, then the next page is not empty. Hence it's the opposite

Oh OK. Probably rename it to `(last_)object_on_page_boundary`, or simply put some comments for it. I thought it meant if the last loop iteration crosses page boundary.

Reply:
Just realized that the accounting for the bulk loading version isn't exactly correct as we don't know if an object is on a page boundary when we've bulk loaded them and they're alive [here](https://github.com/mmtk/mmtk-core/pull/361/files#diff-612b529a28c7592c8dcab06b6364051cf6a553ec527bb40f076ec21def62b0f0R416)

Reply:
Sorry I don't quite understand this. Has this been fixed? If not, can you explain the issue a little bit more? @k-sareen

Reply:
No it hasn't. The issue is what we only know that _something_ is alive, and not necessarily know what is alive and how many bytes is it. In order to check if an object spans a page boundary (i.e. its end has crossed a page boundary), we would have to go through the 128-bits in some fashion to figure out if one of the objects alive spans a page boundary. I don't know what the cost of this would be -- it seems like it would defeat the purpose of the bulk xor

Reply:
If I understand you correctly, the issue you are talking about is that if the 128bit check does not find dead objects, we will not linear scan the region, and we do not know if there will be any live object in the region that lives across the page boundary. In that case, the next page may be incorrectly unset because we think it is empty. Is this correct?

I further checked the use of the page mark, and it seems we never actually check the page mark (`is_page_marked()` is unused). We only set and unset the mark, but never really use the mark. If that's the case, this is not going to be a problem. However, if that's the case, why do we need the page mark at all?

Reply:
> If I understand you correctly, the issue you are talking about is that if the 128bit check does not find dead objects, we will not linear scan the region, and we do not know if there will be any live object in the region that lives across the page boundary. In that case, the next page may be incorrectly unset because we think it is empty. Is this correct?

Yes you are correct.

So for the time being, I am using page mark through an inefficient way, but as outlined [above](https://github.com/mmtk/mmtk-core/pull/361#issuecomment-871969511) we need to figure out a way to do the actual calculation performantly -- maybe something like a SIMD add of 16 bytes at one time.

Reply:
Can you point me to the code where you access the page mark bit? `is_page_marked()` (including the unsafe one) is never used for what I searched. If the page mark is not used, then the issue we talked about doesn't matter as it affects page mark, and page mark is not used. 

I am trying to figure out what needs to be done before this can be merged.

Reply:
I use page mark for my own experiments -- it's not part of this PR as it's not performant enough

Reply:
That makes sense. So we can either keep the page mark metadata if you plan to submit future PRs using it (probably leave a comment for the metadata spec saying it is not used at the moment but you plan to use it later), or we can simply remove it as it is not used.

Reply:
I'll add a comment -- It'll be a bit of effort to remove it and then add it again. I am basically trying to optimize the addition of the metadata as much as possible. It doesn't seem like Rust auto-vectorizes the addition when using references. Maybe we could convert the metadata to an array/`Vec` and then do an add. That might result in SIMD instructions.

Reply:
Done. I'll check why the JikesRVM build fails after dinner. I haven't changed anything other than merge master into my branch.